File: ios/jentis_flutter/Sources/jentis_flutter/JentisFlutterPlugin.swift
import Flutter
import Foundation
import JentisSDK
import UIKit

public class JentisFlutterPlugin: NSObject, FlutterPlugin, JentisApi,
  FlutterStreamHandler
{
  private var eventSink: FlutterEventSink?

  @objc public func onListen(
    withArguments arguments: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    eventSink = events
    DebugURLProtocol.eventSink = eventSink
    URLProtocol.registerClass(DebugURLProtocol.self)
    return nil
  }

  @objc public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    DebugURLProtocol.eventSink = nil
    return nil
  }

  func initialize(config: TrackConfig) throws {
    let environment: JentisSDK.TrackConfig.Environment
    switch config.environment {
    case .live:
      environment = .live
    case .stage:
      environment = .stage
    }

    let protocolString = config.customProtocol.map { proto in
      switch proto {
      case .http: return "http"
      case .https: return "https"
      }
    }

    let config = JentisSDK.TrackConfig(
      trackDomain: config.trackDomain,
      container: config.container,
      environment: environment,
      version: config.version,
      debugCode: config.debugCode,
      sessionTimeoutInSeconds: config.sessionTimeoutInSeconds != nil
        ? Double(config.sessionTimeoutInSeconds!) : nil,
      authorizationToken: config.authorizationToken,
      customProtocol: protocolString,
      enableOfflineTracking: config.enableOfflineTracking
    )

    JentisService.configure(with: config)
  }
  
  // Update configuration for JentisTrackService
  func restartConfig(config: TrackConfig) throws {
    let environment: JentisSDK.TrackConfig.Environment
    switch config.environment {
    case .live:
      environment = .live
    case .stage:
      environment = .stage
    }

    let protocolString = config.customProtocol.map { proto in
      switch proto {
      case .http: return "http"
      case .https: return "https"
      }
    }

    let config = JentisSDK.TrackConfig(
      trackDomain: config.trackDomain,
      container: config.container,
      environment: environment,
      version: config.version,
      debugCode: config.debugCode,
      sessionTimeoutInSeconds: config.sessionTimeoutInSeconds != nil
        ? Double(config.sessionTimeoutInSeconds!) : nil,
      authorizationToken: config.authorizationToken,
      customProtocol: protocolString,
      enableOfflineTracking: config.enableOfflineTracking
    )

    JentisService.restartConfig(config: <#T##TrackConfig#>)
  }

  func setConsents(
    consents: [String: ConsentValue],
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    let vendorConsents = consents.mapValues { value -> ConsentStatus in
      switch value {
      case .allow: return .allow
      case .deny: return .deny
      case .ncm: return .ncm
      }
    }

    Task {
      do {
        try await TrackingService.shared.setConsents(vendorConsents)
        completion(.success(()))
      } catch {
        completion(.failure(error))
      }
    }
  }

  func push(events: [JentisEvent]) throws {
    let eventData = events.map { event -> [String: Any] in
      var dict = [String: Any]()

      if let intAttributes = event.intAttributes {
        dict.merge(intAttributes) { $1 }
      }

      if let boolAttributes = event.boolAttributes {
        dict.merge(boolAttributes) { $1 }
      }

      if let doubleAttributes = event.doubleAttributes {
        dict.merge(doubleAttributes) { $1 }
      }

      if let stringAttributes = event.stringAttributes {
        dict.merge(stringAttributes) { $1 }
      }

      return dict
    }

    for event in eventData {
      try TrackingService.shared.push(event)
    }
  }

  func submit(
    customInitiator: String,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    Task {
      do {
        try await TrackingService.shared.submit(customInitiator)
        completion(.success(()))
      } catch {
        completion(.failure(error))
      }
    }
  }

  func addEnrichment(enrichment: Enrichment) throws {
    try TrackingService.shared.addEnrichment(
      pluginId: enrichment.pluginId,
      arguments: (enrichment.arguments ?? [:]).compactMapValues {
        $0 as? CodableValue
      },
      variables: enrichment.variables ?? []
    )
  }

  func addCustomEnrichment(enrichment: Enrichment) throws {
    try TrackingService.shared.addCustomEnrichment(
      pluginId: enrichment.pluginId,
      arguments: (enrichment.arguments ?? [:]).compactMapValues {
        $0 as? CodableValue
      },
      variables: enrichment.variables ?? []
    )
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = JentisFlutterPlugin()

    JentisApiSetup.setUp(
      binaryMessenger: registrar.messenger(),
      api: instance,
      messageChannelSuffix: ""
    )

    let eventChannel = FlutterEventChannel(
      name: "debug_requests",
      binaryMessenger: registrar.messenger()
    )
    eventChannel.setStreamHandler(instance)
  }
}

class DebugURLProtocol: URLProtocol {
  static var eventSink: FlutterEventSink?

  override class func canInit(with request: URLRequest) -> Bool {
    // Intercept all requests
    return true
  }

  override class func canonicalRequest(for request: URLRequest) -> URLRequest {
    return request
  }

  override func startLoading() {
    let url = request.url?.absoluteString ?? ""
    var bodyString = ""
    if let body = request.httpBody {
      bodyString = String(data: body, encoding: .utf8) ?? ""
    } else if let stream = request.httpBodyStream {
      stream.open()
      let bufferSize = 1024
      var buffer = [UInt8](repeating: 0, count: bufferSize)
      var data = Data()
      while stream.hasBytesAvailable {
        let read = stream.read(&buffer, maxLength: bufferSize)
        if read > 0 {
          data.append(buffer, count: read)
        } else {
          break
        }
      }
      stream.close()
      bodyString = String(data: data, encoding: .utf8) ?? ""
    }

    let debugInfo: [String: Any] = [
      "url": url,
      "body": bodyString,
    ]
    DispatchQueue.main.async {
      DebugURLProtocol.eventSink?(debugInfo)
    }

    // Forward the request to the network
    let session = URLSession(configuration: .default)
    let task = session.dataTask(with: request) { data, response, error in
      if let response = response {
        self.client?.urlProtocol(
          self,
          didReceive: response,
          cacheStoragePolicy: .notAllowed
        )
      }
      if let data = data {
        self.client?.urlProtocol(self, didLoad: data)
      }
      if let error = error {
        self.client?.urlProtocol(self, didFailWithError: error)
      }
      self.client?.urlProtocolDidFinishLoading(self)
    }
    task.resume()
  }

  override func stopLoading() {}
}

File: android/src/main/kotlin/com/jentis/flutter/JentisFlutterPlugin.kt
package com.jentis.flutter

import ConsentValue
import JentisApi
import Enrichment
import JentisEvent
import TrackConfig
import android.app.Application
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.jentis.sdk.jentissdk.JentisTrackService
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import okhttp3.OkHttpClient
import okio.Buffer

/** JentisFlutterPlugin */
class JentisFlutterPlugin : FlutterPlugin, JentisApi, ActivityAware {
    private var context: Context? = null
    private var application: Application? = null
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        JentisApi.setUp(flutterPluginBinding.binaryMessenger, this)

        EventChannel(
            flutterPluginBinding.binaryMessenger, "debug_requests"
        ).setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(
                arguments: Any?, events: EventChannel.EventSink?
            ) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = null
        JentisApi.setUp(binding.binaryMessenger, null)
    }

    // Update configuration for JentisTrackService
    override fun restartConfig(config: TrackConfig) {
        JentisTrackService.getInstance().restartConfig(
            trackDomain = config.trackDomain,
            container = config.container,
            environment = config.environment.name,
            version = config.version,
            debugCode = config.debugCode,
            authToken = config.authorizationToken,
            sessionTimeoutParam = config.sessionTimeoutInSeconds?.toInt(),
            protocol = config.customProtocol?.let { it.toString().lowercase() + "://" },
            enableOfflineTracking = config.enableOfflineTracking,
        )
    }

    override fun initialize(config: TrackConfig) {
        context ?: throw IllegalStateException(
            "Context is not initialized. Make sure to call onAttachedToEngine first."
        )
        application ?: throw IllegalStateException(
            "Application is not initialized. Make sure to set the application in onAttachedToActivity."
        )

        fun getRequestBodyAsString(request: okhttp3.Request): String? {
            val body = request.body ?: return null
            return try {
                val buffer = Buffer()
                body.writeTo(buffer)
                buffer.readUtf8()
            } catch (e: Exception) {
                Log.e(TAG, "The request body could not be read", e)
                null
            }
        }

        val okHttpClient = OkHttpClient.Builder().addInterceptor(
            DebugInterceptor { request ->
                // Send request info to Dart
                Handler(Looper.getMainLooper()).post {
                    eventSink?.success(
                        mapOf(
                            "url" to request.url.toString(),
                            "body" to getRequestBodyAsString(request)
                        )
                    )
                }
            }).build()

        // Initialize tracking service
        JentisTrackService.initialize(context!!, okHttpClient).apply {
            initTracking(
                application = application!!,
                trackDomain = config.trackDomain,
                container = config.container,
                environment = config.environment.name.lowercase(),
                version = config.version,
                debugCode = config.debugCode,
                authToken = config.authorizationToken,
                sessionTimeoutParam = config.sessionTimeoutInSeconds?.toInt(),
                protocol = config.customProtocol?.let { it.toString().lowercase() + "://" },
                enableOfflineTracking = config.enableOfflineTracking,
                offlineTimeout = config.offlineTimeout,
            )
        }
    }

    override fun setConsents(
        consents: Map<String, ConsentValue>, callback: (Result<Unit>) -> Unit
    ) {
        try {
            val vendorConsents = consents.mapValues { entry ->
                when (entry.value) {
                    ConsentValue.ALLOW -> true
                    ConsentValue.DENY -> false
                    ConsentValue.NCM -> "ncm"
                }
            }

            JentisTrackService.getInstance().setConsents(vendorConsents = vendorConsents)
            callback(Result.success(Unit))
        } catch (e: Exception) {
            callback(Result.failure(e))
        }
    }

    override fun push(events: List<JentisEvent>) {
        val eventData = events.map { event ->
            LinkedHashMap<String, Any>().apply {
                event.intAttributes?.let { putAll(it) }
                event.boolAttributes?.let { putAll(it) }
                event.doubleAttributes?.let { putAll(it) }
                event.stringAttributes?.let { putAll(it) }
            }
        }

        JentisTrackService.getInstance().push(eventData)
    }

    override fun submit(customInitiator: String, callback: (Result<Unit>) -> Unit) {
        try {
            JentisTrackService.getInstance().submit(customInitiator)
            callback(Result.success(Unit))
        } catch (e: Exception) {
            callback(Result.failure(e))
        }
    }

    override fun addEnrichment(enrichment: Enrichment) {
        val enrichmentMap =
            mapOf(
                "plugin" to mapOf("pluginId" to enrichment.pluginId), "enrichment" to mapOf(
                    "variablesEnrichment" to (enrichment.variables ?: emptyList())
                ), "args" to (enrichment.arguments?.filterValues { it != null }?.mapValues {
                    it.value as Any
                } ?: emptyMap()))

        JentisTrackService.getInstance().addEnrichment(enrichmentMap)
    }

    override fun addCustomEnrichment(enrichment: Enrichment) {
        val enrichmentMap =
            mapOf(
                "plugin" to mapOf("pluginId" to enrichment.pluginId), "enrichment" to mapOf(
                    "variablesEnrichment" to (enrichment.variables ?: emptyList())
                ), "args" to (enrichment.arguments?.filterValues { it != null }?.mapValues {
                    it.value as Any
                } ?: emptyMap()))

        JentisTrackService.getInstance().addCustomEnrichment(enrichmentMap)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        application = binding.activity.application
    }

    override fun onDetachedFromActivityForConfigChanges() {
        application = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        application = binding.activity.application
    }

    override fun onDetachedFromActivity() {
        application = null
    }

    companion object {
        private const val TAG = "JentisFlutterPlugin"
    }
}
